lua/bufferline/icons.lua

```diff
@@ -16,25 +15,26 @@ local hl_groups = {}
 -- changes, therefore we need to re-define colors for all groups we have
 -- already highlighted.
 local function set_highlights()
-  for i, hl_group in ipairs(hl_groups) do
+  for _, hl_group in ipairs(hl_groups) do
     local icon_hl = hl_group[1]
     local buffer_status = hl_group[2]
-    nvim.command(
-      'hi! ' .. icon_hl .. buffer_status ..
-      ' guifg=' .. get_attr(icon_hl, 'foreground') ..
-      ' guibg=' .. get_attr('Buffer'..buffer_status, 'background')
-    )
+    vim.api.nvim_set_hl(0, icon_hl .. buffer_status, {
+      bg = get_attr('Buffer' .. buffer_status, 'background'),
+      fg = get_attr(icon_hl, 'foreground'),
+    })
   end
 end
 
 
 local function get_icon(buffer_name, filetype, buffer_status)
   if status == false then
-    nvim.command('echohl WarningMsg')
-    nvim.command('echom "barbar: bufferline.icons is set to v:true but \\\"nvim-dev-icons\\\" was not found."')
-    nvim.command('echom "barbar: icons have been disabled. Set bufferline.icons to v:false to disable this message."')
-    nvim.command('echohl None')
-    nvim.command('let g:bufferline.icons = v:false')
+    vim.notify(
+      'barbar: bufferline.icons is set to v:true but \\\"nvim-dev-icons\\\" was not found.' ..
+        '\nbarbar: icons have been disabled. Set `bufferline.icons` to `false` to disable this message.',
+      vim.log.levels.WARN,
+      {title = 'barbar.nvim'}
+    )
+    vim.g.bufferline.icons = false
```

Comment:
This is not going to work IIUC.
In my understanding, `vim.g.bufferline.icons` 
 - reads the vimscript dict `g:bufferline` into a lua table in the luajit VM
 - sets the `icons` field of the lua table to `false`
But this wouldn't update the `g:bufferline.icons` property in the vimscript memory. To be validated though.

Reply:
I'm getting some really weird results here— I didn't expect Neovim to behave this way at _all_. 

For example, running this:

```vim
let g:foo = {'icons': v:false}

echom g:foo
lua vim.pretty_print(vim.g.foo)

lua vim.g.foo.icons = true

echom g:foo
lua vim.pretty_print(vim.g.foo)
```

Outputs this:

```log
{'icons': v:false}                                                                                                                                  
{
  icons = false
}

{'icons': v:false}
{
  icons = false
}
```

* ??? why is `vim.g.foo.icons` still `false`?

Running this:

```vim
let g:foo = {'icons': v:false}

echom g:foo
lua vim.pretty_print(vim.g.foo)

lua vim.g.foo = {icons = true}

echom g:foo
lua vim.pretty_print(vim.g.foo)
```

Outputs what I expected the first time:

```log
{'icons': v:false}                                                                                                                                  
{
  icons = false
}

{'icons': v:true}
{
  icons = true
}
```

Doing the opposite, though:

```vim
lua vim.g.foo = {icons = false}

echom g:foo
lua vim.pretty_print(vim.g.foo)

let g:foo.icons = v:true

echom g:foo
lua vim.pretty_print(vim.g.foo)
```
Works just fine:

```log
{'icons': v:false}                                                                                                                                      
{
  icons = false
}

{'icons': v:true}
{
  icons = true
}
```

Is this a bug? `vim.g.foo = {}` _can_ set a value for `g:foo` in VimScript, but `vim.g.foo.icons` won't update either side.

Reply:
This is 100% expected.
Doing `lua vim.g.foo.icons = true` *never* updates `g:foo`, it updates the (temporary) lua table that is read by `lua vim.g.foo`
If you do for example:
```lua
print(vim.g.foo)
print(vim.g.foo)
```
You're not reading the same table; `vim.g` is converting `g:foo` twice, returning each time a new table.

Reply:
It's unintuitive to me at least that tables returned by `vim.g` don't have metatables assigned which propogate updates in the same way that directly assigning to `vim.g` does. I'll revert this change either way for now, since it won't work